python/ray/tune/syncer.py

```diff
@@ -142,395 +79,408 @@ class SyncConfig:
     """
 
     upload_dir: Optional[str] = None
-    syncer: Optional[str] = "auto"
+    syncer: Optional[Union[str, "Syncer"]] = "auto"
 
     sync_on_checkpoint: bool = True
-    sync_period: int = 300
+    sync_period: int = DEFAULT_SYNC_PERIOD
+
+
+class _BackgroundProcess:
+    def __init__(self, fn: Callable):
+        self._fn = fn
+        self._process = None
+        self._result = {}
+
+    @property
+    def is_running(self):
+        return self._process and self._process.is_alive()
+
+    def start(self, *args, **kwargs):
+        if self.is_running:
+            return False
+
+        self._result = {}
+
+        def entrypoint():
+            try:
+                result = self._fn(*args, **kwargs)
+            except Exception as e:
+                self._result["exception"] = e
+                return
+
+            self._result["result"] = result
+
+        self._process = threading.Thread(target=entrypoint)
+        self._process.start()
+
+    def wait(self):
+        if not self._process:
+            return
+
+        self._process.join()
+        self._process = None
+
+        exception = self._result.get("exception")
+        if exception:
+            raise exception
+
+        result = self._result.get("result")
+
+        self._result = {}
+        return result
 
 
 @DeveloperAPI
-class Syncer:
-    def __init__(self, local_dir: str, remote_dir: str, sync_client: SyncClient = NOOP):
-        """Syncs between two directories with the sync_function.
-
-        Arguments:
-            local_dir: Directory to sync. Uniquely identifies the syncer.
-            remote_dir: Remote directory to sync with.
-            sync_client: Client for syncing between local_dir and
-                remote_dir. Defaults to a Noop.
-        """
-        self._local_dir = os.path.join(local_dir, "") if local_dir else local_dir
-        self._remote_dir = remote_dir
+class Syncer(abc.ABC):
+    """Syncer class for synchronizing data between Ray nodes and external storage.
+
+    This class handles data transfer for two cases:
+
+    1. Synchronizing data from the driver to external storage. This affects
+      experiment-level checkpoints and trial-level checkpoints if no cloud storage
+      is used.
+    2. Synchronizing data from remote trainables to external storage.
+
+    Synchronizing tasks are usually asynchronous and can be awaited using ``wait()``.
```

Comment:
May not be in the scope of this PR as I understand async is current behavior.
However, thinking about the whole story of cloud checkpointing, I keep wonder what should be the right semantics to work with (I also tagged you in my PR on the report API, ptal there).

tldr, there will be cases that user saves to a directory in some session, and then want to checkpoint that directory. Who should be responsible for managing the life cycle of that local directory? 
Option 1: our framework. Then we need to make sure that the local directory won't get blown up by deleting stuff regularly. The API is less intuitive, as we are moving away from with Tune.checkpoint_dir(iteration) kind of API, where iteration is used as the index for sub folder, that Tune keeps track of. 
Option 2: User's responsibility. Seems better with the new report API. However, it may be racing if the semantics of syncing to cloud is "asynchronous".

Reply:
Here is one proposal:

user can always write to "whatever_directory" under the current working directory (which is changed in the beginning)
upon report() returned, our framework makes sure that the move is completed to move "whatever_directory" to "checkpoint_{iter}" directory. The operation should be cheap.
Now, our framework can manage this "checkpoint_{iter}" directory the same way that we do to Tune checkpoints today.
Thus, the restriction on syncing semantics can be removed.
cc @krfricke to cross check with syncer logic.

Reply:
IIUC this will affect how Tune calls the Syncer interface - i.e. which directories it tells the Syncer to sync. From that side any logic would be fine. Let's have the discussion about the actual API in your PR!